addons/crm_addons/controllers/main.py

```diff
@@ -0,0 +1,387 @@
+
+        # Check signature.
+        database_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        signature = hmac.new(database_secret.encode(),
+                             ('%s.%s' % (b64_encoded_header, b64_encoded_payload)).encode(),
+                             sha256).hexdigest()
+        signature_provided = base64.b64decode(
+            b64_encoded_signature).decode()  # decode b64, then decode the resulting binary string
+        if not hmac.compare_digest(signature, signature_provided):
+            return False
+
+        payload_string = base64.b64decode(b64_encoded_payload).decode()
+        payload_json = json.loads(payload_string)
+
+        user_id = payload_json.get('user_id')
+        expiration_time = payload_json.get('time_valid')
+
+        if not user_id or not expiration_time:
+            return False
+
+        if int(expiration_time) < time.time():
+            return False
+
+        user = request.env['res.users'].browse(user_id)
+        if not user:
+            return False
+
+        request.env.user = user.id
+        return True
+
+    @http.route('/crm/addons/login', type='http', auth="none")
+    def web_login(self, redirect=None, **kw):
+        request.params['login_success'] = False
+
+        if not request.uid:
+            request.uid = odoo.SUPERUSER_ID
+
+        values = request.params.copy()
+        try:
+            values['databases'] = http.db_list()
+        except odoo.exceptions.AccessDenied:
+            values['databases'] = None
+
+        if request.httprequest.method == 'POST':
+            try:
+                uid = request.session.authenticate(request.session.db, request.params['login'], request.params['password'])
+                request.params['login_success'] = True
+                code = self._generate_code(uid, int(time.time()) + 30 * 1000)
+                redirect = "%s?success=1&response_type=%s&state=%s&code=%s" % (redirect, kw.get('response_type'), kw.get('state'), code)
+                return http.redirect_with_hash(redirect)
+            except odoo.exceptions.AccessDenied as e:
+                if e.args == odoo.exceptions.AccessDenied().args:
+                    values['error'] = _("Wrong login/password")
+                else:
+                    values['error'] = e.args[0]
+        else:
+            if 'error' in request.params and request.params.get('error') == 'access':
+                values['error'] = _('Only employee can access this database. Please contact the administrator.')
+
+        if 'login' not in values and request.session.get('auth_login'):
+            values['login'] = request.session.get('auth_login')
+
+        if not odoo.tools.config['list_db']:
+            values['disable_database_manager'] = True
+
+        response = request.render('crm_addons.login', values)
+        response.headers['X-Frame-Options'] = 'DENY'
+        return response
+
+    def _generate_code(self, uid, time_valid):
```

Comment:
This is tokens stuff rather than OTP (let alone TOTP). Looks fine to me, though we might eventually want to build a utility to create / sign stuff using the db secret at one point as we're starting to have quite a few of these.